telemetry/derive/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+                _ => false,
+            },
+            _ => false,
+        })
+        .and_then(|arg| match arg {
+            FnArg::Typed(typ) => match *typ.pat.clone() {
+                syn::Pat::Ident(ident) => Some(ident.ident),
+                _ => None,
+            },
+            _ => None,
+        })
+        .ok_or(input)
+}
+
+struct MetricSpecs(Vec<MetricSpec>); // `HashSet` — idiomatic; slow
+
+impl Parse for MetricSpecs {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let vars = Punctuated::<MetricSpec, Comma>::parse_terminated(input)?;
+        Ok(Self(vars.into_iter().collect()))
+    }
+}
+
+struct MetricSpec {
+    #[cfg(feature = "metric-instrumentation")]
+    timing: bool,
+    metric_name: LitStr,
+}
+
+impl Parse for MetricSpec {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let _timing = <syn::Token![+]>::parse(input).is_ok();
+        let metric_name_lit = syn::Lit::parse(input)?;
+        let metric_name = match metric_name_lit {
+            syn::Lit::Str(lit_str) => lit_str,
+            _ => {
+                return Err(syn::Error::new(
+                    proc_macro2::Span::call_site(),
+                    "Must be a string literal. Format `[+]\"name_of_metric\"`.",
+                ))
+            }
+        };
+        Ok(Self {
+            metric_name,
+            #[cfg(feature = "metric-instrumentation")]
+            timing: _timing,
+        })
+    }
+}
+
+impl ToTokens for MetricSpec {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        self.metric_name.to_tokens(tokens)
+    }
+}
+
+/// Macro for instrumenting an `isi`'s `impl execute` to track a given
+/// metric.  To specify a metric, put it as an attribute parameter
+/// inside quotes.
+
+/// This will increment the [`prometheus::IntVec`] metric
+/// corresponding to the literal provided in quotes, with the second
+/// argument being `TOTAL_STR == "total"`. If the execution of the
+/// `Fn`'s body doesn't result in an [`Err`] variant, another metric
+/// with the same first argument and `SUCCESS_STR = "success"` is also
+/// incremented. Thus one can infer the number of rejected
+/// transactions based on this parameter. If necessary, this macro
+/// should be edited to record different [`Err`] variants as different
+/// rejections, so we could (in theory), record the number of
+/// transactions that got rejected because of
+/// e.g. `SignatureCondition` failure.
+///
+/// If you also want to track the execution time of the `isi`, you
+/// should prefix the quoted metric with the `+` symbol.
+///
+/// # Examples
+/// ```rust
+/// use iroha_core::wsv::{World, WorldStateView};
+/// use iroha_telemetry_derive::metrics;
+///
+/// #[metrics(+"test_query", "another_test_query_without_timing")]
+/// fn execute(wsv: &WorldStateView<World>) -> Result<(), ()> {
+///		Ok(())
+/// }
+/// ```
+#[proc_macro_error]
+#[proc_macro_attribute]
+#[allow(clippy::str_to_string)]
```

Comment:
is there need for this lint?

Reply:
Unfortunately, `abort!` seems to use hard-code a `to_string()` instead of `to_owned` and triggers clippy. We need both `str` and `string-to-string` because of the `format!` macros.